Unable to resolve revision of valid tag "@name/subname@v1.0.1"

**This is a copy of issue #204 which is still present.**

<hr>

# Steps to reproduce

main.go

```go
package main

import (
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
)

func main() {
	repo, err := git.PlainOpen(".")
	if err != nil {
		panic(err)
	}

	tagName := "@name/subname@v1.0.1"
	_, err = repo.ResolveRevision(plumbing.Revision(tagName))
	if err != nil {
		panic(err)
	}
}
```

```
$ git tag                                                                                                     
@name/subname@v1.0.1
$ git rev-list -n 1 @name/subname@v1.0.1
c93521b3316ee00c2d8f190b1420ed0adde79f6c          
$ ./main
panic: Revision invalid : reference must be defined once at the beginning                    
```

Workaround

As a workaround, I resolve the tag by reference hash instead of its name.

```go
package main

import (
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
)

func main() {
	repo, err := git.PlainOpen(".")
	if err != nil {
		panic(err)
	}

	tagName := "@name/subname@v1.0.1"
	reference, err := repo.Tag(tagName)
	if err != nil {
		panic(err)
	}

	_, err = repo.ResolveRevision(plumbing.Revision(reference.Hash().String()))
	if err != nil {
		panic(err)
	}
}
```

--- a/internal/revision/parser.go
+++ b/internal/revision/parser.go
@@ -565,6 +565,10 @@ func (p *Parser) parseRef() (Revisioner, error) {
 
 		if endOfRef {
 			p.unscan()
+			if buf == "@" {
+				return Ref("HEAD"), nil
+			}
+
 			return Ref(buf), nil
 		}
 

--- a/internal/revision/parser_test.go
+++ b/internal/revision/parser_test.go
@@ -177,6 +177,13 @@ func (s *ParserSuite) TestParseWithValidExpression() {
 			CaretPath{1},
 			CaretPath{1},
 		},
+		"@scope/package@v1.0.1": []Revisioner{
+			Ref("@scope/package@v1.0.1"),
+		},
+		"@scope/package@v1.0.1~1": []Revisioner{
+			Ref("@scope/package@v1.0.1"),
+			TildePath{1},
+		},
 	}
 
 	for d, expected := range datas {

--- a/internal/revision/scanner.go
+++ b/internal/revision/scanner.go
@@ -88,7 +88,24 @@ func (s *scanner) scan() (token, string, error) {
 	case '-':
 		return minus, string(ch), nil
 	case '@':
-		return at, string(ch), nil
+		next, _, err := s.r.ReadRune()
+		if err == io.EOF || next == zeroRune {
+			return at, string(ch), nil
+		}
+
+		if err != nil {
+			return tokenError, "", err
+		}
+
+		if err := s.r.UnreadRune(); err != nil {
+			return tokenError, "", err
+		}
+
+		if next == '{' {
+			return at, string(ch), nil
+		}
+
+		return word, string(ch), nil
 	case '\\':
 		return aslash, string(ch), nil
 	case '?':

--- a/internal/revision/scanner_test.go
+++ b/internal/revision/scanner_test.go
@@ -133,6 +133,24 @@ func (s *ScannerSuite) TestReadAt() {
 	s.Equal(at, tok)
 }
 
+func (s *ScannerSuite) TestReadAtBeforeBrace() {
+	scanner := newScanner(bytes.NewBufferString("@{1}"))
+	tok, data, err := scanner.scan()
+
+	s.NoError(err)
+	s.Equal("@", data)
+	s.Equal(at, tok)
+}
+
+func (s *ScannerSuite) TestReadAtAsWord() {
+	scanner := newScanner(bytes.NewBufferString("@scope"))
+	tok, data, err := scanner.scan()
+
+	s.NoError(err)
+	s.Equal("@", data)
+	s.Equal(word, tok)
+}
+
 func (s *ScannerSuite) TestReadAntislash() {
 	scanner := newScanner(bytes.NewBufferString("\\"))
 	tok, data, err := scanner.scan()

